src/tui/mod.rs

```diff
@@ -0,0 +1,144 @@
+mod terminal;
+
+use crate::async_buffer_redirect::AsyncBufferRedirect;
+use crate::ShutdownHandle;
+use ansi_to_tui::IntoText;
+use crossterm::event::Event;
+use crossterm::event::EventStream;
+use crossterm::event::KeyCode;
+use crossterm::event::KeyModifiers;
+use crossterm::event::MouseEventKind;
+use miette::miette;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use ratatui::prelude::Buffer;
+use ratatui::prelude::Rect;
+use ratatui::widgets::Paragraph;
+use ratatui::widgets::Widget;
+use ratatui::widgets::Wrap;
+use std::cmp::min;
+use tokio::io::AsyncReadExt;
+use tokio::io::DuplexStream;
+use tokio_stream::StreamExt;
+
+#[derive(Default)]
+struct Tui {
+    quit: bool,
+    scrollback: Vec<u8>,
+    // TODO(evan): Follow output when scrolled to bottom
+    scroll_offset: usize,
+}
+
+/// TODO(evan): Document
+pub async fn run_tui(
+    mut shutdown: ShutdownHandle,
+    mut ghci_reader: DuplexStream,
+) -> miette::Result<()> {
+    let mut tracing_reader = AsyncBufferRedirect::stderr()
+        .into_diagnostic()
+        .wrap_err("Failed to capture stderr")?;
+
+    let mut ghci_buffer = [0; 1024];
+    let mut tracing_buffer = [0; 1024];
+
+    let mut terminal = terminal::enter()?;
+
+    let mut tui = Tui::default();
+
+    let mut event_stream = EventStream::new();
+
+    while !tui.quit {
+        let mut render_result = Ok(());
+        terminal
+            .draw(|frame| {
+                let area = frame.size();
+                let buffer = frame.buffer_mut();
+                render_result = render(&tui, area, buffer);
+            })
+            .into_diagnostic()
+            .wrap_err("Failed to draw to terminal")?;
+        render_result?;
+
+        tokio::select! {
+            _ = shutdown.on_shutdown_requested() => {
+                tui.quit = true;
+            }
+
+            output = ghci_reader.read(&mut ghci_buffer) => {
+                let n = output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from GHCi into TUI buffer")?;
+                if n == 0 {
+                    tui.quit = true;
+                } else {
+                    tui.scrollback.extend(ghci_buffer);
+                    ghci_buffer = [0; 1024];
+                }
+            }
+
+            output = tracing_reader.read(&mut tracing_buffer) => {
+                output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from tracing into TUI buffer")?;
+                tui.scrollback.extend(tracing_buffer);
+                tracing_buffer = [0; 1024];
+            }
```

Comment:
Hmm, I'm skeptical of this... I'm not experienced with `select!` or async cancellation.

I know `AsyncReadExt::read` is cancel safe, but I don't know about the other select arms.

Even if they're all cancel safe, maybe the horrid performance has something to do with creating a new future on every loop or something.

Like maybe instead of something like this:

```rust
let foo: tokio::sync::mpsc::Receiver<usize> = /* ... */;
let bar: tokio::sync::mpsc::Receiver<usize> = /* ... */;
loop {
  select! {
    Some(msg) = foo.recv() => println!("foo received: {msg}"),
    Some(msg) = bar.recv() => println!("bar received: {msg}"),
    else => break,
  }
}
```

...I should have something like this:

```rust
let foo: tokio::sync::mpsc::Receiver<usize> = /* ... */;
let mut foo_fut = foo.recv();
let bar: tokio::sync::mpsc::Receiver<usize> = /* ... */;
let mut bar_fut = bar.recv();
loop {
  select! {
    Some(msg) = foo => {
      println!("foo received: {msg}");
      foo_fut = foo.recv();
    }
    Some(msg) = bar => {
      println!("bar received: {msg}");
      bar_fut = bar.recv();
    }
    else => break,
  }
}
```

...perhaps? Then I'm polling the same future repeatedly, rather than creating a new future on each loop, even if it's semantically the same?

I dunno. Second one's kinda awful lol, I hope that boilerplate isn't required.

Reply:
I think I was wrong about the "re-creating the futures on each loop" concern.

But I am still skeptical of this `select!`, now because I think I should be `tokio::spawn`ing these futures outside of the `loop` (or at least outside of the `select!`) so that the reads can occur in parallel.

Currently, I think they're just happening concurrently on the same thread, because they're obscured from `tokio`'s view; it only sees a single `run_tui` future (or one of `run_tui`'s ancestors). I need to spawn them as new tasks for the executor to be able to schedule them on separate threads.